ntoskrnl/io/pnpmgr/pnpnotify.c

```diff
@@ -19,210 +19,259 @@
+
+    KeReleaseGuardedMutex(&PiNotificationInProgressLock);
+    (Entry->PnpNotificationProc)(NotificationStructure, Entry->Context);
+    KeAcquireGuardedMutex(&PiNotificationInProgressLock);
+}
+
+static
+VOID
+PipProcessNotification(
+    _In_ PVOID NotificationStruct,
+    _In_ IO_NOTIFICATION_EVENT_CATEGORY EventCategory,
+    _In_opt_ PDEVICE_OBJECT DeviceObject)
+{
+    PLIST_ENTRY ListEntry;
+    PPNP_NOTIFY_ENTRY ChangeEntry;
+
+    PAGED_CODE();
+
+    KeAcquireGuardedMutex(&PiNotificationInProgressLock);
+
+    // first, process the main list
+    ListEntry = PiNotifyListHead.Flink;
+    while (ListEntry != &PiNotifyListHead)
     {
         ChangeEntry = CONTAINING_RECORD(ListEntry, PNP_NOTIFY_ENTRY, PnpNotifyList);
-        CallCurrentEntry = FALSE;
 
-        if (ChangeEntry->EventCategory != EventCategory)
-        {
-            ListEntry = ListEntry->Flink;
-            continue;
-        }
+        ListEntry = ListEntry->Flink; // in case this entry would be deleted
 
-        switch (EventCategory)
-        {
-            case EventCategoryDeviceInterfaceChange:
-            {
-                if (RtlCompareUnicodeString(&ChangeEntry->Guid, &GuidString, FALSE) == 0)
-                {
-                    CallCurrentEntry = TRUE;
-                }
-                break;
-            }
-            case EventCategoryHardwareProfileChange:
-            {
-                CallCurrentEntry = TRUE;
-                break;
-            }
-            case EventCategoryTargetDeviceChange:
-            {
-                Status = IoGetRelatedTargetDevice(ChangeEntry->FileObject, &EntryDeviceObject);
-                if (NT_SUCCESS(Status))
-                {
-                    if (DeviceObject == EntryDeviceObject)
-                    {
-                        if (Event == &GUID_PNP_CUSTOM_NOTIFICATION)
-                        {
-                            ((PTARGET_DEVICE_CUSTOM_NOTIFICATION)NotificationStructure)->FileObject = ChangeEntry->FileObject;
-                        }
-                        else
-                        {
-                            ((PTARGET_DEVICE_REMOVAL_NOTIFICATION)NotificationStructure)->FileObject = ChangeEntry->FileObject;
-                        }
-                        CallCurrentEntry = TRUE;
-                    }
-                }
-                break;
-            }
-            default:
-            {
-                DPRINT1("IopNotifyPlugPlayNotification(): unknown EventCategory 0x%x UNIMPLEMENTED\n", EventCategory);
-                break;
-            }
-        }
+        PipProcessSingleNotification(ChangeEntry, EventCategory, NotificationStruct, DeviceObject);
+    }
+
+    // Now process added elements (if any)
+    ListEntry = PiAddedNotifyListHead.Flink;
+    while ((ListEntry = ExInterlockedRemoveHeadList(&PiAddedNotifyListHead, &PiNotifyAdditionLock)))
+    {
+        // append the entry to PiNotifyListHead while we're holding PiNotificationInProgressLock
```

Comment:
```suggestion
        /* Append the entry to PiNotifyListHead while we're holding PiNotificationInProgressLock */
```
OK the code is already self-explaining, so no need to say that?